framework/src/meshgenerators/RefineSidesetGenerator.C

```diff
@@ -0,0 +1,132 @@
+      true,
+      "Toggles whether neighboring level one elements should be refined or not. Defaults to true");
+  MultiMooseEnum boundary_side("primary secondary both", "both");
+  params.addParam<MultiMooseEnum>("boundary_side",
+                                  boundary_side,
+                                  "Whether the generator should refine itself(primary), its "
+                                  "neighbors(secondary), or itself and its neighbors(both)");
+
+  return params;
+}
+
+RefineSidesetGenerator::RefineSidesetGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input(getMesh("input")),
+    _boundaries(getParam<std::vector<BoundaryName>>("boundaries")),
+    _refinement(getParam<std::vector<int>>("refinement")),
+    _enable_neighbor_refinement(getParam<bool>("enable_neighbor_refinement")),
+    _boundary_side(getParam<MultiMooseEnum>("boundary_side"))
+{
+}
+
+std::unique_ptr<MeshBase>
+RefineSidesetGenerator::generate()
+{
+  // Get the list of boundary ids from the boundary names
+  const auto boundary_ids = MooseMeshUtils::getBoundaryIDs(
+      *_input, getParam<std::vector<BoundaryName>>("boundaries"), true);
+
+  // Check that the boundary ids/names exist in the mesh
+  for (std::size_t i = 0; i < boundary_ids.size(); ++i)
+    if (boundary_ids[i] == Moose::INVALID_BOUNDARY_ID)
+    {
+      if (isParamValid("_boundaries"))
+        paramError("boundaries",
+                   "The boundary '",
+                   getParam<std::vector<BoundaryName>>("boundaries")[i],
+                   "' was not found within the mesh");
+    }
+
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+  int max = *std::max_element(_refinement.begin(), _refinement.end());
+
+  return recursive_refine(boundary_ids, mesh, _refinement, max);
+}
+
+std::unique_ptr<MeshBase>
+RefineSidesetGenerator::recursive_refine(std::vector<boundary_id_type> boundary_ids,
+                                         std::unique_ptr<MeshBase> & mesh,
+                                         std::vector<int> refinement,
+                                         int max,
+                                         int ref_step)
+{
+  // If the refinement step has reached the largest value in the _refinement array, return the mesh,
+  // as we are done.
+  if (ref_step == max)
+    return dynamic_pointer_cast<MeshBase>(mesh);
+  mesh->prepare_for_use();
+  // BoundaryInfo boundInfo = mesh->get_boundary_info();
+  std::vector<std::tuple<dof_id_type, unsigned short int, boundary_id_type>> sideList =
+      mesh->get_boundary_info().build_active_side_list();
+  for (std::size_t i = 0; i < boundary_ids.size(); i++)
+  {
+    if (refinement[i] > 0 && refinement[i] > ref_step)
+    {
+      for (std::tuple<dof_id_type, unsigned short int, boundary_id_type> tuple : sideList)
+      {
+        if (std::get<2>(tuple) == boundary_ids[i])
+        {
+          Elem * elem = mesh->query_elem_ptr(std::get<0>(tuple));
```

Comment:
```suggestion
          Elem * elem = mesh->elem_ptr(std::get<0>(tuple));
```

I think you want this to error if the element doesn't exist, no?